### PR TITLE
[Issue 8340] [pulsar-testclient] Fix to support to specify topics and subscriptions

### LINKED ISSUE #8340: [pulsar-testclient]Provide options to specify existing topics when allowAutoTopicCreation is disable and using multiple topics
**Is your enhancement request related to a problem? Please describe.**
Some users may encounter the following issues when using multiple topics with option `--num-topics` but `allowAutoTopicCreation` is disable:
```
$ bin/pulsar-perf consume persistent://public/default/test -s sub -st Exclusive -t 2
Exception in thread "main" java.util.concurrent.ExecutionException: org.apache.pulsar.client.api.PulsarClientException$TopicDoesNotExistException: Topic does not exist
        at java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:357)
        at java.util.concurrent.CompletableFuture.get(CompletableFuture.java:1895)
        at org.apache.pulsar.testclient.PerformanceConsumer.main(PerformanceConsumer.java:343)
Caused by: org.apache.pulsar.client.api.PulsarClientException$TopicDoesNotExistException: Topic does not exist
        at org.apache.pulsar.client.impl.ClientCnx.getPulsarClientException(ClientCnx.java:1003)
        at org.apache.pulsar.client.impl.ClientCnx.handleError(ClientCnx.java:610)
        at org.apache.pulsar.common.protocol.PulsarDecoder.channelRead(PulsarDecoder.java:170)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
```
or using multiple subscriptions with option `--num-consumers` but `allowAutoSubscriptionCreation` is disable:
```
$ bin/pulsar-perf consume persistent://public/default/test -s sub -st Exclusive -n 2
[pulsar-client-io-1-8] WARN  org.apache.pulsar.client.impl.ConnectionHandler - [persistent://public/default/test-partition-3] [sub-1] Could not get connection to broker: Subscription does not exist -- Will try again in 0.1 s
[pulsar-client-io-1-7] WARN  org.apache.pulsar.client.impl.ConnectionHandler - [persistent://public/default/test-partition-4] [sub-1] Could not get connection to broker: Subscription does not exist -- Will try again in 0.1 s
[pulsar-client-io-1-3] WARN  org.apache.pulsar.client.impl.ConnectionHandler - [persistent://public/default/test-partition-2] [sub-0] Could not get connection to broker: Subscription does not exist -- Will try again in 0.1 s
[pulsar-client-io-1-4] WARN  org.apache.pulsar.client.impl.ConnectionHandler - [persistent://public/default/test-partition-1] [sub-0] Could not get connection to broker: Subscription does not exist -- Will try again in 0.1 s
```

**Describe the solution you'd like**
Add some options to specify existing topics or subscriptions to meet the above scenario.

### PR DESCRIPTION
**Pull request by murong00**
### Motivation

Fixes #8340

### Modifications

- Fix option `main parameter` for CLI `pulsar-perf produce/consume/read` to specify multi topics.
- Add option `--subscriptions` for CLI `pulsar-perf` consume to specify multi subscriptions.
- Deprecate and hide the option `--subscriber-name`.
- Modify the corresponding doc.

### CODE REVIEW

**Comment by murong00** [comment]
@codelipenghui Please help to take a review about this when you are available.

File: pulsar-testclient/src/main/java/org/apache/pulsar/testclient/PerformanceConsumer.java [hunk 1]
```diff
@@ -72,8 +71,8 @@
         @Parameter(names = { "--conf-file" }, description = "Configuration file")
         public String confFile;
 
-        @Parameter(description = "persistent://prop/ns/my-topic", required = true)
-        public List<String> topic;
+        @Parameter(description = "A list of topics to consume from (e.g. persistent://prop/ns/topic1 persistent://prop/ns/topic2)", required = true)
+        public List<String> topics;
```
Comment (eolivelli) [inline pulsar-testclient/src/main/java/org/apache/pulsar/testclient/PerformanceConsumer.java:75]: are we breaking compatibility with the old format ?

File: pulsar-testclient/src/main/java/org/apache/pulsar/testclient/PerformanceProducer.java [hunk 2]
```diff
@@ -109,7 +108,7 @@
         @Parameter(names = { "-s", "--size" }, description = "Message size (bytes)")
         public int msgSize = 1024;
 
-        @Parameter(names = { "-t", "--num-topic" }, description = "Number of topics")
+        @Parameter(names = { "-t", "--num-topics" }, description = "Number of topics")
```
Comment (eolivelli) [inline pulsar-testclient/src/main/java/org/apache/pulsar/testclient/PerformanceProducer.java:111]: are we breaking compatibility with the previous version ?

**Review by eolivelli — CHANGES_REQUESTED** [review]
I appreciate this work, happy to see that


Please do not break compatibility with the previous signature.
We have scripts that are used to work with many different versions of pulsar (just by changing the docker image name).

IMHO in general it is fine to add new feature, but we should not change the behaviour if it is not strictly nececessary

File: pulsar-testclient/src/main/java/org/apache/pulsar/testclient/PerformanceConsumer.java (hunk 1 above)
Comment (murong00) [reply pulsar-testclient/src/main/java/org/apache/pulsar/testclient/PerformanceConsumer.java]: Thanks for pointing this out, just keep the old format.

File: pulsar-testclient/src/main/java/org/apache/pulsar/testclient/PerformanceProducer.java (hunk 2 above)
Comment (murong00) [reply pulsar-testclient/src/main/java/org/apache/pulsar/testclient/PerformanceProducer.java]: Done, use the previous version.

**Review by codelipenghui — APPROVED** [review]

**Review by zymap — APPROVED** [review]
Please avoid changing the document and the code with the same PR. Pulsar's document does not need to cherry-pick to the release branch and this may introduce some unnecessary work when we cherry-pick this to the target branch.

**Comment by eolivelli** [comment]
@zymap 
This patch broke compatibility with the previous version.
I have tools that used the useful feature, that is to auto create many topics on demand and automatically

I left comments and you committed the patch anyway.

This is not the right way to behave IMHO.

Can you please revert this patch ?

Otherwise I will be happy to send a patch that at least restores the auto-topic creation behaviour

`bin/pulsar-perf produce -t 10 test`

should create 10 topics, now it does not work anymore
```
The size of topics list should be equal to --num-topic
Usage: pulsar-perf produce [options] persistent://prop/ns/my-topic
  Options:
```

**Comment by codelipenghui** [comment]
Currently, we don't have any tests that cover the test client, we'd better add tests to avoid the broken.

**Comment by eolivelli** [comment]
@codelipenghui I agree.
But I added a comment and asked explicitly to not change the behaviour.

IMHO This is not about tests, it is about taking care of comments of the users in the community.

In my experience in OSS projects, and especially in the ASF,
if anyone "request changes" or anyway has a perplexity regarding a change,
it is necessary to understand the problem deeper before moving forward with committing the patch.
 
Think about any other user/contributor that tries to comment and express his opinion on a change, but the owners of the project ignore his comments and go forward.......
how can the newcomer feel ?
This is not a good behaviour indeed.

Please take this into consideration for the next time.

The is no hurry in committing a patch in Apache projects,
but we must ensure that everybody feels well in the community
:-)

**Comment by zymap** [comment]
@eolivelli I am sorry for my mistake and the inconvenience. 
I saw your all suggested changes already addressed by him and the command arguments look good for compatibility. 
I will revert the change from branch 2.7 and we can fix the issue at master.

**Comment by eolivelli** [comment]
@zymap no problem.
Thanks.

I can send a patch on Monday.

I also have other improvements for pulsar-perf to contribute as well  ;)

**Comment by codelipenghui** [comment]
@eolivelli 

> But I added a comment and asked explicitly to not change the behaviour.
> IMHO This is not about tests, it is about taking care of comments of the users in the community.

You have mentioned 2 points and the author has fixed them. I'm not sure why to make you feel we are not taking care of your comments. If some other places break the behavior, we'd better point out? Otherwise, other reviewers may not understand where the break is.

**Comment by murong00** [comment]
@eolivelli I'm so sorry that I misunderstood  your comment  about the `breaking compatibility`. Actually, I have discussed this with penghui on the issue page:
image

Both you and penghui are right about the behaviour. I think we can just fix this by adding an option to cover the compatibility. I can send a patch if you needed @eolivelli, sorry again about the mistake.

**Comment by eolivelli** [comment]
@murong00 
don't worry at all!
let's fix this together.

I appreciate that you are volunteering to follow up. Thank you very much.

I really would like to see that same behaviour as before, otherwise the behaviour of the tool is different from one version to another and you cannot know which version of the tool you are using.
I am maintaining a suite of tests that are to be run against several different versions of Pulsar.
It is fine to ADD new features, but NOT to change the behaviour.

So out-of-the-box it would be great that the behaviour does not change.
So adding a new "compatibility mode" flag won't work for me, because I cannot pass that flag to old versions.

I was going to start writing a patch in order to fallback to 2.7 mode in case you pass only one topic to the command.
`pulsar-perf produce -n 10 test`
here we see that the user is selecting only one topic and asking for 10...so we can expand the list of topics to
test-1, test-2, test-3....

this way we are 100% compatible with previous behaviour and your new behaviour works as well

The same is for `pulsar-perf consume` please

I will be happy to test your patch

**Comment by murong00** [comment]
@eolivelli Thanks for your tips, I will send a patch later to cover this.

**Pull request merged**